Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2798 @@
+    NSString *docID = [self stringReferenceObjectForObjectID:moid];
+
+    NSEntityDescription *entity = [mo entity];
+    NSDictionary *propDic = [entity propertiesByName];
+
+    NSMutableDictionary *props = [NSMutableDictionary dictionary];
+    NSMutableDictionary *blobStore = [NSMutableDictionary dictionary];
+    NSDictionary *changed = [mo changedValues];
+
+    for (NSString *name in changed) {
+        NSDictionary *enc = nil;
+        id prop = propDic[name];
+        if ([prop isTransient]) {
+            continue;
+        }
+        if ([prop isKindOfClass:[NSAttributeDescription class]]) {
+            NSAttributeDescription *att = prop;
+            enc =
+                [self encodeAttribute:att withObject:changed[name] blobStore:blobStore error:&err];
+        } else if ([prop isKindOfClass:[NSRelationshipDescription class]]) {
+            NSRelationshipDescription *rel = prop;
+            enc = [self encodeRelation:rel withObject:changed[name] error:&err];
+        } else {
+            oops(@"bad prop?");
+        }
+        [props addEntriesFromDictionary:enc];
+    }
+
+    CDTDocumentRevision *oldRev = [self.datastore getDocumentWithId:docID error:&err];
+    if (!oldRev) {
+        if (error) *error = err;
+        return NO;
+    }
+
+    // TODO: version HACK
+    NSString *oldVersion = oldRev.body[kCDTISObjectVersionKey];
+    uint64_t version = [oldVersion longLongValue];
+    ++version;
+    NSNumber *v = [NSNumber numberWithUnsignedLongLong:version];
+    props[kCDTISObjectVersionKey] = [v stringValue];
+
+    CDTMutableDocumentRevision *upRev = [oldRev mutableCopy];
+
+    // update attachments first
+    if ([blobStore count]) {
+        [upRev.attachments addEntriesFromDictionary:blobStore];
+    }
+
+    // delete all changed properties, in case they are being removed.
+    [upRev.body removeObjectsForKeys:[props allKeys]];
+    [upRev.body addEntriesFromDictionary:props];
+
+    CDTDocumentRevision *upedRev = [self.datastore updateDocumentFromRevision:upRev error:&err];
+    if (!upedRev) {
+        if (error) *error = err;
+        return NO;
+    }
+
+    if ([blobStore count]) {
+        if (![self.datastore compactWithError:&err]) {
+            CDTLogWarn(CDTREPLICATION_LOG_CONTEXT, @"%@: datastore compact failed: %@", kCDTISType,
+                       err);
+        }
+    }
+
+    if (CDTISReadItBack) {
+        /**
+         *  See CDTISReadItBack
+         */
+        upedRev = [self.datastore getDocumentWithId:upRev.docId error:&err];
+        if (!upedRev) {
+            // Always oops!
+            oops(@"ReadItBack: error: %@", err);
+        }
+        // Always oops
+        if ([upedRev.body count] == 0) oops(@"empty save");
+    }
+
+    return YES;
+}
+
+/**
+ *  Delete a managed object from the database
+ *
+ *  @param mo    Managed Object
+ *  @param error Error
+ *
+ *  @return YES/NO
+ */
+- (BOOL)deleteManagedObject:(NSManagedObject *)mo error:(NSError **)error
```

Comment:
Does CoreData handle cascading deletes above the incremental store layer?  You implementation of delete seems to depend on this.  It might be worth adding a note in the block comment on this assumption/dependency.